util/collections/src/garbage/atomic/counted.rs

```diff
@@ -0,0 +1,159 @@
+use std::sync::Arc;
+
+/// An atomic Garbage Collected counter.
+/// The total count accumulated can be garbage collected periodically using the `gc` method.
+/// Counts by slot are stored in a ring buffer. So, reused slots will also be garbage collected when they are reassigned.
+/// The reusage of slots does not remove the need to call `gc` periodically, as slots which are not reused would not be garbage collected, causing the count to drift.
+#[derive(Debug, Clone)]
+pub struct GcCounter {
+	/// The number of some unit time a value is valid for.
+	value_ttl: Duration,
+	/// The duration of a garbage collection slot in some unit time.
+	gc_slot_duration: Duration,
+	/// The array of atomic counters for value lifetimes, where each entry represents a slot with a timestamp and count.
+	value_lifetimes: Arc<Vec<(AtomicU64, AtomicU64)>>,
+	/// The number of slots calculated as value_ttl / gc_slot_duration.
+	num_slots: u64,
+}
+
+impl GcCounter {
+	/// Creates a new GcCounter with a specified garbage collection slot duration.
+	pub fn new(value_ttl: Duration, gc_slot_duration: Duration) -> Self {
+		let num_slots = value_ttl.get() / gc_slot_duration.get();
+		let value_lifetimes =
+			Arc::new((0..num_slots).map(|_| (AtomicU64::new(0), AtomicU64::new(0))).collect());
+		GcCounter { value_ttl, gc_slot_duration, value_lifetimes, num_slots }
+	}
+
+	/// Decrements the value, saturating over non-zero slots.
+	pub fn decrement(&self, mut amount: u64) {
+		for (_, count) in self.value_lifetimes.iter() {
+			// Use `fetch_update` to perform a safe, atomic update
+			let result = count.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |current_count| {
+				if current_count == 0 {
+					None // Stop if the count is already zero
+				} else if current_count >= amount {
+					Some(current_count - amount) // Deduct the full amount
+				} else {
+					// Otherwise, subtract what we can and let `amount` carry the rest
+					amount -= current_count;
+					Some(0)
+				}
+			});
+
+			// If the update was successful or if the slot is zero, we can move on
+			if result.is_ok() || amount == 0 {
+				break;
+			}
+
+			// Stop early if the remaining amount has been fully decremented
+			if amount == 0 {
+				break;
+			}
+		}
+	}
+
+	/// Increments the value in a specific slot.
+	pub fn increment(&self, current_time: u64, amount: u64) {
+		let slot_timestamp = current_time / self.gc_slot_duration.get();
+		let slot = slot_timestamp % self.num_slots;
+		let (active_slot_timestamp, count) = &self.value_lifetimes[slot as usize];
+
+		// Atomically check and set the timestamp if it doesn't match current_time
+		let active_slot = active_slot_timestamp.load(Ordering::Relaxed);
+		if active_slot == slot {
+			// Same timestamp, increment count
+			count.fetch_add(amount, Ordering::SeqCst);
+		} else {
+			// Different timestamp, reset slot for new period
+			active_slot_timestamp.store(slot_timestamp, Ordering::SeqCst);
+			count.store(amount, Ordering::SeqCst);
+		}
```

Comment:
I think there is a race condition: if two threads call `increment` concurrently at the same timestamp, and the slot has not been updated, they will both set each own `amount` and one of them gets overwritten.

Reply:
In other words, these should always set the same slot where we would have a race. Update would provide a weak guarantee in the same way that decrement is handled.

Reply:
> Perhaps we can check that the slot pushed is not less than the other slot?

Slots are modulo the vector length, perhaps you mean the timestamps. This is not good enough because you want to check if the value has not been changed between the load and the store as well. Maybe you need to use `compare_exchange_weak` instead.

Reply:
`fetch_update` is more complete...

Reply:
But, it doesn't quite work here.

Reply:
I think I may have it.